mathics/builtin/numbers/calculus.py

```diff
@@ -812,6 +812,193 @@ def to_sympy(self, expr, **kwargs):
+    >> Reduce[x y == 0,{x, y}]
+     = x == 0  ||   y == 0 
+    >> Reduce[{x y == 0, y == 4},{x, y}]
+     = y == 4 && x == 0
+    >> Reduce[{x y == 0, y > 4},{x, y}]
+     =  x == 0 && y > 4
+    >> Reduce[x*y>0,{x,y}]
+     = (x < 0 && y < 0) || (x > 0 && y > 0)
+    >> Reduce[{x*y > 0, y>0},{x,y}]
+     =  x > 0 && y > 0
+    
+    >> Reduce[a<b,a]
+     =  b ∈ Reals && a < b
+    """
+    
+    messages = {
+        "eqf": "`1` is not a well-formed equation.",
+        "svars": 'Equations may not give solutions for all "solve" variables.',
+    }
+
+    rules = {
+        "Reduce[eqs_, vars_, Complexes]": "Reduce[eqs, vars]",
+       
+    }
+
+    summary_text = "find generic solutions for variables, in logical expression form"
+    
+    
+    def apply(self, eqs, vars, evaluation):
+        "Reduce[eqs_, vars_]"
+        
+        vars_original = vars
+        head_name = vars.get_head_name()
+        if head_name == "System`List":
+            vars = vars.leaves
+        else:
+            vars = [vars]
+        for var in vars:
+            if (
+                (var.is_atom() and not var.is_symbol())
+                or head_name in ("System`Plus", "System`Times", "System`Power")  # noqa
+                or constant & var.get_attributes(evaluation.definitions)
+            ):
+
+                evaluation.message("Solve", "ivar", vars_original)
+                return
+        eqs_original = eqs
+        if eqs.get_head_name() in ("System`List", "System`And"):
+            eqs = eqs.leaves
+        else:
+            eqs = [eqs]
+        sympy_eqs = []
+        sympy_denoms = []
+        for eq in eqs:
+            if eq is SymbolTrue:
+                pass
+            elif eq is SymbolFalse:
+                return Expression(SymbolList)
+            elif not eq.has_form("Equal", 2):
+                if any(eq.has_form(operator,2) for operator in ["Less","LessEqual","Greater","GreaterEqual"]):
+                    eq=eq.to_sympy()
+                else:     
+                    return evaluation.message("Solve", "eqf", eqs_original)
+            else:
+                left, right = eq.leaves
+                left = left.to_sympy()
+                right = right.to_sympy()
+                if left is None or right is None:
+                    return
+                eq = left - right
+                eq = sympy.together(eq)
+                eq = sympy.cancel(eq)
+                numer, denom = eq.as_numer_denom()
+                sympy_denoms.append(denom)
+            sympy_eqs.append(eq)
+
+        vars_sympy = [var.to_sympy() for var in vars]
+        if None in vars_sympy:
+            return
+
+        # delete unused variables to avoid SymPy's
+        # PolynomialError: Not a zero-dimensional system
+        # in e.g. Solve[x^2==1&&z^2==-1,{x,y,z}]
+        all_vars = vars[:]
+        all_vars_sympy = vars_sympy[:]
+        vars = []
+        vars_sympy = []
+        for var, var_sympy in zip(all_vars, all_vars_sympy):
```

Comment:
It would be nice to understand and explain in a comment what we are doing in this chunk of code.